packages/protobuf/src/extension-accessor.ts

```diff
@@ -0,0 +1,125 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { Message } from "./message.js";
+import type { BinaryReadOptions, BinaryWriteOptions } from "./binary-format.js";
+import { assert } from "./private/assert.js";
+import {
+  createExtensionContainer,
+  filterUnknownFields,
+} from "./private/extensions.js";
+import type { Extension } from "./extension.js";
+
+/**
+ * Retrieve an extension value from a message.
+ *
+ * The function never returns undefined. Use hasExtension() to check whether an
+ * extension is set. If the extension is not set, this function returns the
+ * default value (if one was specified in the protobuf source), or the zero value
+ * (for example `0` for numeric types, `[]` for repeated extension fields, and
+ * an empty message instance for message fields).
+ */
+export function getExtension<E extends Message<E>, V>(
+  message: E,
+  extension: Extension<E, V>,
+  options?: Partial<BinaryReadOptions>,
+): V {
+  const opt = extension.runtime.bin.makeReadOptions(options);
+  assert(!extension.field.oneof); // oneof is not allowed for extensions
+  assert(extension.field.kind != "map"); // maps are not allowed for extensions
+  const ufs = filterUnknownFields(
+    message.getType().runtime.bin.listUnknownFields(message),
+    extension.field,
+  );
+  const [container, get] = createExtensionContainer(extension);
+  for (const uf of ufs) {
+    extension.runtime.bin.readField(
+      container,
+      opt.readerFactory(uf.data),
+      extension.field,
+      uf.wireType,
+      opt,
+    );
+  }
+  return get();
+}
+
+/**
+ * Set an extension value on a message. If the message already has a value for
+ * this extension, the value is replaced.
+ */
+export function setExtension<E extends Message<E>, V>(
+  message: E,
+  extension: Extension<E, V>,
+  value: V,
+  options?: Partial<BinaryReadOptions & BinaryWriteOptions>,
+): void {
+  const readOpt = extension.runtime.bin.makeReadOptions(options);
+  const writeOpt = extension.runtime.bin.makeWriteOptions(options);
+  if (hasExtension(message, extension)) {
+    const ufs = message
+      .getType()
+      .runtime.bin.listUnknownFields(message)
+      .filter((uf) => uf.no != extension.field.no);
+    message.getType().runtime.bin.discardUnknownFields(message);
+    for (const uf of ufs) {
+      message
+        .getType()
+        .runtime.bin.onUnknownField(message, uf.no, uf.wireType, uf.data);
+    }
+  }
+  const writer = writeOpt.writerFactory();
+  extension.runtime.bin.writeField(extension.field, value, writer, writeOpt);
+  const reader = readOpt.readerFactory(writer.finish());
+  while (reader.pos < reader.len) {
+    const [no, wireType] = reader.tag();
+    const data = reader.skip(wireType);
```

Comment:
If wire type is "start group", how does this know the tag number to look for in the closing "end group" tag?

Reply:
It doesn't. Method `skip` returns the bytes until (and including) the "end group" tag. It recursively skips nested groups. We're iterating over tags we just created ourselves, so we should be fine in this case.

There is a check for matching end group field number when we parse a group [here](https://github.com/bufbuild/protobuf-es/blob/5ee26a1f465af8a2ea703a9aa9ac21758a87ccbb/packages/protobuf/src/private/binary-format-common.ts#L191). IMO it would be better to check in method `skip`, so that we verify unknown fields for matching group start/end tags as well, but I'm not sure it's worth it right now, as we're already moving a lot of parts.

Reply:
Ack. Maybe a TODO in `reader.skip`?